FLT/HIMExperiments/module_topology.lean

```diff
@@ -206,13 +203,159 @@ noncomputable def LinearMap.neg (M : Type*) [AddCommGroup M] [Module A M] :
+@ContinuousAdd P (Module.topology A) _ := by
+  apply @ContinuousAdd.mk _ (topology A)
+  rw [Module.prod_canonical A]
+  exact Module.continuous_linear A (LinearMap.add A P)
+
+variable [Module.Finite A M] [Module.Free A M] [Module.Finite A N] [Module.Free A N]
+
+
+-- I really shoouldn't start with iA instead of topology A here.
+-- But I need it like this in the next proof and don't know how to rewrite it there
+-- Sorry for leaving it like this...
+instance Module.instContinuousSMul : @ContinuousSMul A M _ (topology A) (topology A) := by
+  rw [← Module.topology_self]
+  apply @ContinuousSMul.mk A M _ _ (topology A)
+  let ι := Free.ChooseBasisIndex A M
+  have b : Basis ι A M := by
+    simp only [ι]
+    exact Free.chooseBasis A M
+  have : (fun (p : A × M) ↦ p.1 • p.2) = (fun p ↦ ∑ i : ι, p.1 • b.equivFun p.2 i • b i) := by
+    ext x
+    rw [← Finset.smul_sum, Basis.sum_equivFun]
+  rw [this]
+  apply @continuous_finset_sum _ _ _ (@instTopologicalSpaceProd _  _ _ (topology A)) (topology A) _ _ _ _
+  intro i _
+  have : (fun (a : A × M) ↦ a.1 • b.equivFun a.2 i • b i) = (fun a ↦ (a.1 * b.equivFun a.2 i) • b i) := by
+    ext
+    rw [← mul_smul]
+  rw [this]
+  have : (fun a ↦ (a.1 * b.equivFun a.2 i) • b i) = (fun (a : A) ↦ a • b i) ∘ (fun (m : A × A) ↦ m.1 * m.2) ∘ (fun (m : A × M) ↦ (m.1, b.equivFun m.2 i)) := by
+    ext a
+    simp
+  rw [this]
+  apply @Continuous.comp _ _ _ (@instTopologicalSpaceProd _ _ _ (topology A)) _ (topology A) _ _
+  · have : (fun a ↦ a • b i) = (LinearMap.smul₁ A M (b i)) := by
+      unfold LinearMap.smul₁
+      simp
+    rw [this]
+    nth_rewrite 1 [Module.topology_self A]
+    apply Module.continuous_linear A (LinearMap.smul₁ A M (b i))
+  apply @Continuous.comp _ _ _ (@instTopologicalSpaceProd _ _ _ (topology A)) _ _ _ _
+  exact continuous_mul
+  apply @Continuous.prod_map _ _ _ _ _ _ _ (topology A) (fun m ↦ m) (fun m ↦ b.equivFun m i)
+  exact continuous_id
+  have : (fun m ↦ b.equivFun m i) = LinearMap.basis₁ A M ι b i := by
+    unfold LinearMap.basis₁
+    simp
+  rw [this]
+  nth_rewrite 2 [Module.topology_self A]
+  apply Module.continuous_linear A (LinearMap.basis₁ A M ι b i)
+
+
+
+-- I assume this is true! Lots of things like this seem to be true.
+lemma Module.continuous_bilinear {P : Type*} [AddCommGroup P] [Module A P] [Module.Finite A P] [Module.Free A P]
+    (f : M →ₗ[A] N →ₗ[A] P) :
+    @Continuous (M × N) P (Module.topology A) (Module.topology A) (fun mn ↦ f mn.1 mn.2) := by
+  let ι := Free.ChooseBasisIndex A M
+  let κ := Free.ChooseBasisIndex A N
+  have b : Basis ι A M := by --how can I name the index ι without doing all this?
+    simp only [ι]
+    exact Free.chooseBasis A M
+  have d : Basis κ A N := by
+    simp only [κ]
+    exact Free.chooseBasis A N
+  rw [← prod_canonical]
+  have : (fun (mn : M × N) ↦ (f mn.1) mn.2) =
+    (fun (mn : M × N) ↦  (∑ k : κ, (∑ i : ι, d.equivFun mn.2 k • b.equivFun mn.1 i • f (b i)) (d k))) := by
```

Comment:
```suggestion
    (fun (mn : M × N) ↦ (∑ k : κ, (∑ i : ι, d.repr mn.2 k • b.repr mn.1 i • f (b i)) (d k))) := by
```
I think all the equivFuns should be changed to reprs throughout the entire proof. `simp` is going to do that anyway the moment you give it the chance. Don't fight the simplifier.